README
NAME
    HTTP::Tiny::Mech - Wrap a WWW::Mechanize instance in an HTTP::Tiny
    compatible interface.

VERSION
    version 1.001002

SYNOPSIS
      # Get something that expects an HTTP::Tiny instance
      # to work with HTTP::Mechanize under the hood.
      #
      my $thing => ThingThatExpectsHTTPTiny->new(
        ua => HTTP::Tiny::Mech->new()
      );

      # Get something that expects HTTP::Tiny
      # to work via WWW::Mechanize::Cached
      #
      my $thing => ThingThatExpectsHTTPTiny->new(
        ua => HTTP::Tiny::Mech->new(
          mechua => WWW::Mechanize::Cached->new( )
        );
      );

DESCRIPTION
    This code is somewhat poorly documented, and highly experimental.

    Its the result of a quick bit of hacking to get "MetaCPAN::API" working
    faster via the "WWW::Mechanize::Cached" module ( and gaining cache
    persistence via "CHI" )

    It works so far for this purpose.

    At present, only "get" and "request" are implemented, and all other
    calls fall through to a native "HTTP::Tiny".

ATTRIBUTES
  "mechua"
    This class provides one non-standard parameter not in HTTP::Tiny,
    "mechua", which is normally an autovivified "WWW::Mechanize" instance.

    You may override this parameter if you want to provide a custom instance
    of a "WWW::Mechanize" class.

WRAPPED METHODS
  get
    Interface should be the same as it is with "get" in HTTP::Tiny.

  request
    Interface should be the same as it is with "request" in HTTP::Tiny

AUTHOR
    Kent Fredric <kentnl@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Kent Fredric <kentnl@cpan.org>.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Release history for HTTP-Tiny-Mech

1.001003 2017-03-10T13:29:10Z 509fe97

1.001002 2017-03-10T13:22:58Z 91e962a
 - Bugfix: Avoid test failures with -Ddefault_inc_excludes_dot
 - Minor test refactoring
 - Added compat for older Test::More

 [Dependencies::Stats]
 - Dependencies changed since 1.001001, see misc/*.deps* for details
 - configure: +1 (recommends: ↑1)
 - develop: +14 ↑4 -3 (recommends: +1, suggests: ↑2)
 - test: +2 ↓1 (recommends: ↑1 ↓1)

1.001001 2014-08-15T06:11:34Z
 [00 Trivial]
 - no code changes.
 - metdata updates
 - whitespace adjustments.
 - test tweaks.

 [Dependencies::Stats]
 - Dependencies changed since 1.001000, see misc/*.deps* for details
 - configure: ↓1
 - develop: +1 (suggests: ↑1)
 - test: +1 -3 (recommends: ↑1 -1)

 [Metadata]
 - 1x Contributor is now recognised as contributor instead of an author

 [Misc]
 - POD moved around so generated source has less whitespace at the top.

 [Tests]
 - Test::ReportPrereqs was updated automatically, and is now more useful.

1.001000 2014-07-17T15:22:19Z
 [00 Major]
 - Recoded construction to work without Class::Tiny, as the logic is much more difficult since 1.000
 - Test Coverage increased to 100%
 - Misc packaging cleanups

 [Dependencies::Stats]
 - Dependencies changed since 1.000000, see misc/*.deps* for details
 - develop: +3 -3 (suggests: +2 -1)
 - runtime: -1
 - test: +1

1.000000 2014-05-10T02:41:46Z
 [00 Minor]
 - Mostly a packaging changes, not interesting code changes, but downstream vendors who package by hand may need to pay
   attention.

 [01 Tooling]
 - Now uses ExtUtils::MakeMaker instead of Module::Build
 - Now defaults to Minimum requirements and "latest" are only recommended ( Previously it required very recent versions
   of toolchain things needlessly )

 [02 Version Scheme]
 - Moved from x.y.z to x.yyyzzz
 - This is the primary reason for the 1.x on the box.

 [Dependencies::Aggregate]
 - 5.008009 : 64 → 61
 - 5.010001 : 48 → 45
 - 5.018002 : ? → 25
 - 5.019000 : 29 → ?
 - 5.019009 : ? → 28

 [Dependencies::Stats]
 - Dependencies changed since 0.2.1, see misc/*.deps* for details
 - build: -1
 - configure: +1 -1 (recommends: +1)
 - develop: +7 ↑1 -9
 - test: +4 ↓1 (recommends: +4)

0.2.1 2013-11-24T00:55:35Z
 [Dependencies::Noteworthy]
 - Needs HTTP::Tiny 0.022 for tests.

 [Dependencies::Stats]
 - Dependencies changed since 0.2.0, see misc/*.deps* for details
 - runtime: ↑1
 - test: +1

0.2.0 2013-11-24T00:18:47Z
 [00 Major]
 - Internals refactored in terms of Class::Tiny, removing the need for Moose to do basically nothing.

 [Dependencies::Noteworthy]
 - use Class::Tiny
 - no longer uses Moose and MooseX::NonMoose

 [Dependencies::Stats]
 - Dependencies changed since 0.1.4, see misc/*.deps* for details
 - develop: ↑1
 - runtime: +2 -2

 [Documentation]
 - Changes retroactively tided.

 [Tests]
 - Basic constructor tests added.

0.1.4 2013-11-21T07:32:56Z
 [00 Maint release]

 [Dependencies::Noteworthy]
 - use Module::Build 0.4202
 - use Test::More 1.001002

 [Dependencies::Stats]
 - Dependencies changed since 0.1.3, see misc/*.deps* for details
 - build: ↑1
 - configure: ↑1
 - develop: +45 -1 (recommends: -1, suggests: +1 -1)
 - test: ↑1 -2

 [Documentation]
 - Update ©
 - UTF-8 encoding

 [Meta]
 - Issues to Github

 [Packaging]
 - utilise test_requires
 - utilise new fallback code

 [Tests]
 - Use new Compile::PerFile
 - Regen report versions to hide develop reqs
 - Regen CPAN Changes test
 - Regen Kwalitee test

0.1.3 2012-11-08T14:06:30Z
 [Bugs]
 - Resolves github #1 "Adjust get() to bridge HTTP::Tiny and WWW::Mechanize APIs", reported by Pedro Melo, with fix by
   Pedro Melo.
 - Resolves RT#80674 , essentially github #1, reported by Diogo Franco

 [Dependencies::Noteworthy]
 - perl = 5.006
 - use Module::Build 0.4003
 - use Test::More 0.98

 [Dependencies::Stats]
 - Dependencies changed since 0.1.2, see misc/*.deps* for details
 - build: ↑1
 - configure: ↑1
 - develop: +5 (recommends: ↑1, suggests: ↑1)
 - runtime: +1
 - test: ↑1

 [Documentation]
 - Added Pedro Melo to author list

 [Tests]
 - regenerated t/00-compile.t
 - regenerated t/000-report-versions-tiny.t
 - added xt/release/minimum-version.t

0.1.2 2011-10-17T09:42:07Z
 [Internals]
 - Query the Request Object for success code instead of passing 1.

0.1.1 2011-10-17T03:48:43Z
 [Packaging]
 - Bloody UTF8 Encoding.

0.1.0 2011-10-17T03:44:31Z
 - First version.


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.009.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Wrap a WWW::Mechanize instance in an HTTP::Tiny compatible interface.",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "HTTP-Tiny-Mech",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "HTTP::Tiny::Mech",
  "PREREQ_PM" => {
    "HTTP::Request" => 0,
    "HTTP::Tiny" => "0.022",
    "WWW::Mechanize" => 0,
    "parent" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "HTTP::Response" => 0,
    "HTTP::Tiny" => "0.022",
    "Test::More" => 0,
    "lib" => 0
  },
  "VERSION" => "1.001003",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t"
  }
);


my %FallbackPrereqs = (
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "HTTP::Request" => 0,
  "HTTP::Response" => 0,
  "HTTP::Tiny" => "0.022",
  "Test::More" => 0,
  "WWW::Mechanize" => 0,
  "lib" => 0,
  "parent" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/HTTP/Tiny/Mech.pm
use 5.006;    # pragmas, our
use strict;
use warnings;

package HTTP::Tiny::Mech;

our $VERSION = '1.001003';

# ABSTRACT: Wrap a WWW::Mechanize instance in an HTTP::Tiny compatible interface.

# AUTHORITY

use parent 'HTTP::Tiny';

sub new {
  my ( $self, %args ) = @_;
  my ( $mechua, $has_mechua );
  if ( exists $args{mechua} ) {
    $has_mechua = 1;
    $mechua     = delete $args{mechua};
  }
  my $instance = $self->SUPER::new(%args);
  if ($has_mechua) {
    $instance->{mechua} = $mechua;
  }
  return bless $instance, $self;
}

## no critic (Subroutines::RequireArgUnpacking)
sub mechua {
  my ( $self, $new_mechua, $has_new_mechua ) = ( $_[0], $_[1], @_ > 1 );
  if ($has_new_mechua) {
    $self->{mechua} = $new_mechua;
  }
  return $self->{mechua} if exists $self->{mechua};
  require WWW::Mechanize;
  return ( $self->{mechua} = WWW::Mechanize->new() );
}
## use critic

=attr C<mechua>

This class provides one non-standard parameter not in HTTP::Tiny, C<mechua>, which
is normally an autovivified C<WWW::Mechanize> instance.

You may override this parameter if you want to provide a custom instance of a C<WWW::Mechanize> class.

=cut

sub _unwrap_response {
  my ( undef, $response ) = @_;
  return {
    status  => $response->code,
    reason  => $response->message,
    headers => $response->headers,
    success => $response->is_success,
    content => $response->decoded_content,
  };
}

sub _wrap_request {
  my ( undef, $method, $uri, $opts ) = @_;
  require HTTP::Request;
  my $req = HTTP::Request->new( $method, $uri );
  $req->headers( $opts->{headers} ) if $opts->{headers};
  $req->content( $opts->{content} ) if $opts->{content};
  return $req;
}

=head1 WRAPPED METHODS

=head2 get

Interface should be the same as it is with L<HTTP::Tiny/get>.

=cut

sub get {
  my ( $self, $uri, $opts ) = @_;
  return $self->_unwrap_response( $self->mechua->get( $uri, ( $opts ? %{$opts} : () ) ) );
}

=head2 request

Interface should be the same as it is with L<HTTP::Tiny/request>

=cut

sub request {
  my ( $self, @request ) = @_;
  my $req      = $self->_wrap_request(@request);
  my $response = $self->mechua->request($req);
  return $self->_unwrap_response($response);
}

=head1 SYNOPSIS

  # Get something that expects an HTTP::Tiny instance
  # to work with HTTP::Mechanize under the hood.
  #
  my $thing => ThingThatExpectsHTTPTiny->new(
    ua => HTTP::Tiny::Mech->new()
  );

  # Get something that expects HTTP::Tiny
  # to work via WWW::Mechanize::Cached
  #
  my $thing => ThingThatExpectsHTTPTiny->new(
    ua => HTTP::Tiny::Mech->new(
      mechua => WWW::Mechanize::Cached->new( )
    );
  );

=cut

=head1 DESCRIPTION

This code is somewhat poorly documented, and highly experimental.

Its the result of a quick bit of hacking to get L<< C<MetaCPAN::API>|MetaCPAN::API >> working faster
via the L<< C<WWW::Mechanize::Cached>|WWW::Mechanize::Cached >> module ( and gaining cache persistence via
L<< C<CHI>|CHI >> )

It works so far for this purpose.

At present, only L</get> and L</request> are implemented, and all other calls
fall through to a native L<< C<HTTP::Tiny>|HTTP::Tiny >>.

=cut

1;


maint/perlcritic.rc.gen.pl
#!/usr/bin/env perl
## no critic (Modules::RequireVersionVar)

# FILENAME: bundle_to_ini.pl
# CREATED: 02/06/14 01:48:56 by Kent Fredric (kentnl) <kentfredric@gmail.com>
# ABSTRACT: Write an INI file from a bundle

use 5.008;    #utf8
use strict;
use warnings;
use utf8;

use Carp qw( croak carp );
use Perl::Critic::ProfileCompiler::Util qw( create_bundle );

## no critic (ErrorHandling::RequireUseOfExceptions)
my $bundle = create_bundle('Example::Author::KENTNL');
$bundle->configure;

my @stopwords = (
  qw(
    Melo autovivified
    )
);
for my $wordlist (@stopwords) {
  $bundle->add_or_append_policy_field( 'Documentation::PodSpelling' => ( 'stop_words' => $wordlist ) );
}

$bundle->remove_policy('CodeLayout::RequireUseUTF8');

#$bundle->remove_policy('ErrorHandling::RequireCarping');
#$bundle->remove_policy('NamingConventions::Capitalization');

my $inf = $bundle->actionlist->get_inflated;

my $config = $inf->apply_config;

{
  open my $rcfile, '>', './perlcritic.rc' or croak 'Cant open perlcritic.rc';
  $rcfile->print( $config->as_ini, "\n" );
  close $rcfile or croak 'Something fubared closing perlcritic.rc';
}
my $deps = $inf->own_deps;
{
  open my $depsfile, '>', './perlcritic.deps' or croak 'Cant open perlcritic.deps';
  for my $key ( sort keys %{$deps} ) {
    $depsfile->printf( "%s~%s\n", $key, $deps->{$key} );
    *STDERR->printf( "%s => %s\n", $key, $deps->{$key} );
  }
  close $depsfile or carp 'Something fubared closing perlcritic.deps';
}



maint-travis-ci/branch_reset.pl
#!/usr/bin/env perl
use strict;
use warnings;

use FindBin;
use lib "$FindBin::Bin/lib";
use tools;

if ( not env_true('TRAVIS') ) {
  diag('Is not running under travis!');
  exit 1;
}
diag("Resetting branch to \e[32m$ENV{TRAVIS_BRANCH}\e[0m @ \e[33m$ENV{TRAVIS_COMMIT}\e[0m");
git( 'checkout', $ENV{TRAVIS_BRANCH} );
git( 'reset', '--hard', $ENV{TRAVIS_COMMIT} );
my $goodtag;
do {
  my $output = capture_stdout {
    git( 'describe', '--tags', '--abbrev=0', $ENV{TRAVIS_BRANCH} );
  };
  ($goodtag) = split /\n/, $output;
};
diag("TIP Version tag is \e[32m$goodtag\e[0m");
my %good_tags;
do {
  my $output = capture_stdout {
    git( 'log', '--simplify-by-decoration', '--pretty=format:%d' );
  };
  for my $line ( split /\n/, $output ) {
    if ( $line =~ /\(tag:\s+(.*)\)/ ) {
      my $tag = $1;
      diag("Good tag: \e[32m$tag\e[0m");
      $good_tags{$tag} = 1;
    }
    else {
      diag("Line not matched regexp: <\e[31m$line\e[0m>");
    }
  }
};
do {
  my $output = capture_stdout {
    git('tag');
  };
  for my $line ( split /\n/, $output ) {
    next if $good_tags{$line};
    diag("Bad tag: \e[31m$line\e[0m");
    git( 'tag', '-d', $line );
  }
};


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.4202"
  },
  "configure_requires" => {
    "Module::Build" => "0.4202"
  },
  "dist_abstract" => "Wrap a WWW::Mechanize instance in an HTTP::Tiny compatible interface.",
  "dist_author" => [
    "Kent Fredric <kentnl\@cpan.org>",
    "Pedro Melo <melo\@simplicidade.org>"
  ],
  "dist_name" => "HTTP-Tiny-Mech",
  "dist_version" => "0.2.1",
  "license" => "perl",
  "module_name" => "HTTP::Tiny::Mech",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Class::Tiny" => 0,
    "HTTP::Request" => 0,
    "HTTP::Tiny" => "0.022",
    "WWW::Mechanize" => 0,
    "parent" => 0,
    "perl" => "5.006",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => [],
  "test_requires" => {
    "HTTP::Tiny" => "0.022",
    "Test::More" => "1.001002"
  }
);


my %fallback_build_requires = (
  "HTTP::Tiny" => "0.022",
  "Module::Build" => "0.4202",
  "Test::More" => "1.001002"
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;
